### PR TITLE
Feature/99 Toggleable sidebar

### LINKED ISSUE #99: Make sidebar closable

### PR DESCRIPTION
**Pull request by moetazaneta**
Closes #99

### CODE REVIEW

**Comment by NightOfNights** [comment]
Don't forget about specifying task number in commit names, about assigning pr to yourself and requesting review from us 😡

File: src/utils/useLocalStorage.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { useState } from 'react';
+
+/** Sync state to local storage so that it persists through a page refresh. */
+export function useLocalStorage<T>(key: string, initialValue: T): [T, ((value: T) => void)] {
```
Comment (NightOfNights) [inline src/utils/useLocalStorage.ts]: I get it that you copy-pasted this hook from somewhere, but, please, at least make documentation style consistent with the rest of our app

File: src/utils/useLocalStorage.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      // Save state
+      setStoredValue(value);
+
+      // Save to local storage
+      if (typeof window !== 'undefined') {
+        window.localStorage.setItem(key, JSON.stringify(value));
+      }
+    } catch (error: unknown) {
+      // A more advanced implementation would handle the error case
+    }
```
Comment (NightOfNights) [inline src/utils/useLocalStorage.ts]: I suppose it's not advanced implementation :(

File: src/utils/useLocalStorage.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+export function useLocalStorage<T>(key: string, initialValue: T): [T, ((value: T) => void)] {
+  // State to store our value
+  // Pass initial state function to useState so logic is only executed once
+  const [storedValue, setStoredValue] = useState<T>(() => {
+    if (typeof window === 'undefined') {
+      return initialValue;
+    }
+    try {
+      // Get from local storage by key
+      const item = window.localStorage.getItem(key);
```
Comment (NightOfNights) [inline src/utils/useLocalStorage.ts]: As far as I understand this checks are needed for [SSR](https://github.com/vercel/next.js/discussions/19911) solutions. I think we can omit them

File: src/components/Sidebar/Sidebar.tsx [hunk 4]
```diff
@@ -1,52 +1,70 @@
...
+import { Heading, Link, IconButton, HStack } from '@chakra-ui/react';
 import { NavLink } from 'react-router-dom';
 import cls from './Sidebar.module.css';
 import { data, SidebarElementType } from './data';
+import { Icon } from '../Icon/Icon';
+
+interface Props {
+
+  /** Envoken on sidebar visibility toggle. */
+  onSiderbarHide: () => unknown;
```
Comment (NightOfNights) [inline src/components/Sidebar/Sidebar.tsx]: I think handle functions should return `void` in the most of cases

File: src/components/Header/Header.tsx [hunk 5]
```diff
@@ -1,12 +1,34 @@
+import { Heading, HStack, IconButton } from '@chakra-ui/react';
 import React, { VFC } from 'react';
 import { Icon } from '../Icon/Icon';
 import cls from './Header.module.css';
 
+interface Props {
+
+  /** Whether logo and sidebar toggle button are shown. */
+  showLogo: boolean;
```
Comment (NightOfNights) [inline src/components/Header/Header.tsx]: use `is` or `has` prefixes for boolean property names

File: src/layouts/DefaultLayout/DefaultLayout.tsx [hunk 6]
```diff
@@ -2,19 +2,28 @@ import React, { VFC } from 'react';
...
+        <div className={cls.sidebar}>
+          <Sidebar onSiderbarHide={() => serSiderbarVisibility(false)} />
+        </div>
+      )}
+      <div className={`${cls.container} ${isSidebarVisible ? cls['with-sidebar'] : ''}`}>
+        <Header showLogo={!isSidebarVisible} onSiderbarShow={() => serSiderbarVisibility(true)} />
+        <div className={cls.content}>
+          <Outlet />
+        </div>
+      </div>
```
Comment (NightOfNights) [inline src/layouts/DefaultLayout/DefaultLayout.tsx]: Let's stick to one approach and use chakra `Box` component instead of `div`

**Review by NightOfNights — CHANGES_REQUESTED** [review]

File: src/components/Header/Header.tsx [hunk 7]
```diff
@@ -1,12 +1,34 @@
...
 /**
  * TODO: Implement header funtionality.
  */
-export const Header: VFC = () => (
+export const Header: VFC<Props> = ({ showLogo, onSiderbarShow }) => (
   <header className={cls.header}>
+    {showLogo && (
+      <HStack justifyContent="space-between" alignItems="center">
+        <IconButton
+          onClick={() => onSiderbarShow()}
```
Comment (Moltenship) [inline src/components/Header/Header.tsx]: Do you really need to pass this as anonymous function? I think that just `onClick={onSidebarShow}` will work

File: src/layouts/DefaultLayout/DefaultLayout.tsx [hunk 8]
```diff
@@ -2,19 +2,28 @@ import React, { VFC } from 'react';
...
-export const DefaultLayout: VFC = () => (
-  <div className={cls.layout}>
-    <div className={cls.sidebar}>
-      <Sidebar />
-    </div>
-    <div className={cls.content}>
-      <Header />
-      <Outlet />
+export const DefaultLayout: VFC = () => {
+  const [isSidebarVisible, serSiderbarVisibility] = useLocalStorage('vndbre:sidebar-visible', true);
```
Comment (Moltenship) [inline src/layouts/DefaultLayout/DefaultLayout.tsx]: Typo

File: src/layouts/DefaultLayout/DefaultLayout.tsx [hunk 9]
```diff
@@ -2,19 +2,28 @@ import React, { VFC } from 'react';
...
+
+  return (
+    <div className={cls.layout}>
+      {isSidebarVisible && (
+        <div className={cls.sidebar}>
+          <Sidebar onSiderbarHide={() => serSiderbarVisibility(false)} />
```
Comment (Moltenship) [inline src/layouts/DefaultLayout/DefaultLayout.tsx]: It is better to use `useCallback` for such cases.

Comment (Moltenship) [reply src/layouts/DefaultLayout/DefaultLayout.tsx]:
```ts
const closeSidebar = useCallback(() => {
  setSidebarVisibility(false);
}, []);

**Review by Moltenship — CHANGES_REQUESTED** [review]

**Comment by moetazaneta** [comment]
> Don't forget about specifying task number in commit names, about assigning pr to yourself and requesting review from us rage

@NightOfNights Sorry 👉👈
Won't happen again!

File: src/utils/useLocalStorage.ts (hunk 1 above)
Comment (moetazaneta) [reply src/utils/useLocalStorage.ts]: I think we should adopt some library with better implementation.

File: src/utils/useLocalStorage.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Sync state to local storage so that it persists through a page refresh.
+ * @param key Local store key.
+ * @param initialValue Initial value, returned on errors.
+ */
+export function useLocalStorage<T>(key: string, initialValue: T): [T, ((value: T) => void)] {
+  // Pass initial state function to useState so logic is only executed once
+  const [storedValue, setStoredValue] = useState<T>(() => {
+    const item = window.localStorage.getItem(key);
```
Comment (NightOfNights) [inline src/utils/useLocalStorage.ts]: You can access `localStorage` without `window`. `window` is needed for SSR solutions

File: src/utils/useLocalStorage.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    return item ? (JSON.parse(item) as T) : initialValue;
```
Comment (NightOfNights) [inline src/utils/useLocalStorage.ts]: `item` check is implicit

File: src/layouts/DefaultLayout/DefaultLayout.tsx [hunk 12]
```diff
@@ -1,20 +1,33 @@
...
-      <Sidebar />
-    </div>
-    <div className={cls.content}>
-      <Header />
-      <Outlet />
-    </div>
-  </div>
-);
+export const DefaultLayout: VFC = () => {
+  const [isSidebarVisible, setSiderbarVisibility] = useLocalStorage('vndbre:sidebar-visible', true);
```
Comment (NightOfNights) [inline src/layouts/DefaultLayout/DefaultLayout.tsx]: Please, move out `localStorage` keys to constants

**Review by NightOfNights — APPROVED** [review]

**Pull request merged**